File: abduction/alp_dx.pl
:- module(alp_dx, [
    explain_obs/2,        % explain_obs(+ObsList, -Delta)
    explain_obs_k/3,      % explain_obs_k(+ObsList, +K, -Delta)
    assume_obs/1,         % assume_obs(+ObsList)
    clear_obs/0,          % clear previously asserted obs/1
    add_ic/1,             % add_ic(+ForbiddenLiteralsList)
    clear_ics/0,          % clear integrity constraints
    abducible/1,          % declare abducible functor/arity
    load_demo/0
]).

:- dynamic abducible/1.
:- dynamic ic/1.
:- dynamic obs/1.
:- dynamic causes/2.

% ------------------------ Abductive core (minimal by #abduced) ------------------------

explain_obs(ObsList, Delta) :-
    between(0, inf, K),
    explain_obs_k(ObsList, K, Delta), !.

explain_obs_k(ObsList, K, Delta) :-
    goals_from_obs(ObsList, Goal),
    prove(Goal, [], D0, K),
    valid_ic(D0),
    sort(D0, Delta).

goals_from_obs([], true).
goals_from_obs([S|Ss], (obs(S), Gs)) :- goals_from_obs(Ss, Gs).

prove(true, D, D, _) :- !.
prove((A,B), D0, D, K) :- !,
    prove(A, D0, D1, K),
    prove(B, D1, D,  K).

% Observation must be covered by some abduced (or abducible) disease via causes/2
prove(obs(S), D0, D, K) :- !,
    ( member(Dz, D0), Dz = disease(Dx), causes(Dx, S) ->
        D = D0
    ; causes(Dx, S),
      prove(disease(Dx), D0, D, K)
    ).

% Non-abducible: prove from program
prove(G, D, D, _) :-
    \+ is_abducible(G),
    call(G), !.

% Abduce an abducible literal if budget allows and ICs not violated
prove(G, D0, D, K) :-
    is_abducible(G),
    K > 0,
    \+ memberchk_eq(G, D0),
    K1 is K - 1,
    D1 = [G|D0],
    \+ violates_ic(D1),
    D = D1.

is_abducible(G) :- functor(G, F, N), abducible(F/N).
memberchk_eq(X, [Y|_]) :- X =@= Y, !.
memberchk_eq(X, [_|T]) :- memberchk_eq(X, T).

% ------------------------ Integrity constraints ------------------------

add_ic(Pat) :- must_be(list, Pat), assertz(ic(Pat)).
clear_ics :- retractall(ic(_)).

valid_ic(D) :- \+ violates_ic(D).
violates_ic(D) :- ic(P), subset_match(P, D), !.

subset_match([], _).
subset_match([P|Ps], D) :-
    member(E, D),
    P =@= E,
    subset_match(Ps, D).

% ------------------------ Observation helpers ------------------------

assume_obs(ObsList) :- maplist(assertz_obs, ObsList).
assertz_obs(S) :- (obs(S) -> true ; assertz(obs(S))).
clear_obs :- retractall(obs(_)).

% ------------------------ Iterative deepening ------------------------
between(L, _, L).
between(L, inf, X) :- succ(L, L1), between(L1, inf, X).

% ------------------------ Diagnosis KB: Gout vs RA ------------------------

% Only diseases are abducible
:- initialization(init_abducibles).
init_abducibles :-
    retractall(abducible(_)),
    assertz(abducible(disease/1)).

% Gout (acute inflammatory, often monoarticular)
causes(gout, monoarticular_big_toe_pain).
causes(gout, sudden_onset_night).
causes(gout, severe_joint_pain).
causes(gout, redness).
causes(gout, swelling).
causes(gout, warmth).
causes(gout, uric_acid_high).
causes(gout, response_to_colchicine).

% Rheumatoid arthritis (chronic, symmetric small joints)
causes(rheumatoid_arthritis, symmetric_small_joints_pain).
causes(rheumatoid_arthritis, morning_stiffness_gt_60min).
causes(rheumatoid_arthritis, swelling).
causes(rheumatoid_arthritis, elevated_esr_crp).
causes(rheumatoid_arthritis, rheumatoid_factor_positive).
causes(rheumatoid_arthritis, anti_ccp_positive).
causes(rheumatoid_arthritis, chronic_course).

% Integrity constraints (optional): forbid co-abducing gout and RA
:- initialization(init_demo_ics).
init_demo_ics :-
    clear_ics,
    add_ic([disease(gout), disease(rheumatoid_arthritis)]).

% Convenience: reset observations + ICs
load_demo :-
    clear_obs,
    init_demo_ics.

% ---------- LP consistency ----------

% lp_consistent(+HList)
% True if HList is consistent with the KB (no contradictions).
lp_consistent(HList) :-
    % Example: just assume all are consistent for now
    % Replace with your real LP checks (e.g., no conflicts, no negation).
    \+ inconsistent(HList).

% You could define `inconsistent/1` in terms of your rules if needed.

% ---------- PLP probability ----------

% plp_prob(+HList, -P)
% Returns joint probability P(HList).
plp_prob([], 1.0).
plp_prob([H|T], P) :-
    plp_prob_single(H, PH),
    plp_prob(T, PT),
    P is PH * PT.

% plp_prob_single(+H, -P)
% Define probabilities for each hypothesis literal (toy example).
plp_prob_single(disease(gout), 0.3).
plp_prob_single(disease(rheumatoid_arthritis), 0.2).
plp_prob_single(rare_genetic_factor(_), 0.01).
plp_prob_single(_, 0.1).

% ---------- Argumentation defeat ----------

% defeat_strength(+Claim, -D)
% D in [0,1], 0 = undefeated, 1 = fully defeated.
% Here we just stub it; integrate your argumentation engine instead.
defeat_strength(Claim, 0.0) :-
    % if no attacking argument found:
    \+ attacked(Claim), !.
defeat_strength(Claim, 0.7) :-
    attacked(Claim), !.

% attacked/1 can be defined by your argumentation framework.

% ---------- Abduction cost ----------

% abduction_cost(+HList, -Cost)
% Basic MDL-like cost: number of hypotheses or more complex formula.
abduction_cost(HList, Cost) :-
    length(HList, N),
    Cost is N.

% ---------- Conditional probability P(EDU|H) ----------

% p_edu_given_h(+ClaimAtom, +HList, -P)
% Probability that ClaimAtom holds given HList.
p_edu_given_h(Claim, HList, 0.9) :-
    % Example: if Claim is entailed by HList, give high probability
    entails(HList, Claim), !.
p_edu_given_h(_, _, 0.1).

% entails/2 must be defined using your ALP / LP rules.

File: abduction/alp_dx.py
from __future__ import annotations
from typing import Sequence, List
from pathlib import Path
import os

from pyswip import Prolog, Functor, Variable, Query

HERE = Path(__file__).resolve().parent
PL_FILE = HERE / "alp_dx.pl"

class AlpDiagnosis:
    def __init__(self, pl_path: str | os.PathLike = PL_FILE):
        pl_path = Path(pl_path)
        if not pl_path.exists():
            raise FileNotFoundError(f"Prolog file not found: {pl_path}")
        self.prolog = Prolog()
        self._consult(str(pl_path))

    def _consult(self, path: str):
        # normalize path for SWI on Windows
        p = path.replace("\\", "/")
        list(self.prolog.query(f'consult("{p}")'))

    def load_demo(self) -> None:
        self._call0("load_demo")

    def clear_obs(self) -> None:
        self._call0("clear_obs")

    def assume_obs(self, symptoms: Sequence[str]) -> None:
        self._call_list("assume_obs", symptoms)

    def clear_ics(self) -> None:
        self._call0("clear_ics")

    def add_ic(self, literals: Sequence[str]) -> None:
        if not literals:
            return
        body = "[" + ",".join(literals) + "]"
        self._call0(f"add_ic({body})")

    def explain_obs(self, symptoms: Sequence[str]) -> List[str]:
        return next(self.explain_obs_all(symptoms), [])

    def explain_obs_all(self, symptoms: Sequence[str]):
        syms = "[" + ",".join(self._to_atom(s) for s in symptoms) + "]"
        q = f"explain_obs({syms}, Delta)"
        for sol in self._safe_query(q):
            yield self._read_delta(sol)

    def explain_obs_k(self, symptoms: Sequence[str], k: int) -> List[str]:
        return next(self.explain_obs_k_all(symptoms, k), [])

    def explain_obs_k_all(self, symptoms: Sequence[str], k: int):
        syms = "[" + ",".join(self._to_atom(s) for s in symptoms) + "]"
        q = f"explain_obs_k({syms}, {int(k)}, Delta)"
        for sol in self._safe_query(q):
            yield self._read_delta(sol)

    # ---------------- utils ----------------

    def _call0(self, goal: str) -> bool:
        return bool(list(self.prolog.query(goal, maxresult=1)))

    def _call_list(self, pred: str, items: Sequence[str]) -> bool:
        prolog_list = "[" + ",".join(self._to_atom(s) for s in items) + "]"
        return self._call0(f"{pred}({prolog_list})")

    @staticmethod
    def _to_atom(x: str) -> str:
        s = str(x).strip().lower().replace(" ", "_")
        if not s or any(c for c in s if not (c.isalnum() or c == "_")) or s[0].isdigit():
            return f"'{s}'"
        return s

    def _read_delta(self, sol) -> List[str]:
        delta_py = sol.get("Delta")
        if isinstance(delta_py, list):
            return [str(t) for t in delta_py]
        # fallback: term_to_atom(Delta, S)
        varS = Variable()
        t2a = Functor("term_to_atom", 2)
        with self._open_query(t2a, [delta_py, varS]) as q:
            if q.nextSolution():
                return self._parse_list_atom(str(varS.value))
        return []

    @staticmethod
    def _parse_list_atom(s: str) -> List[str]:
        s = s.strip()
        if not (s.startswith("[") and s.endswith("]")):
            return [s]
        inner = s[1:-1].strip()
        return [p.strip() for p in inner.split(",")] if inner else []

    def _safe_query(self, q: str, maxresult: int | None = None):
        try:
            it = self.prolog.query(q) if maxresult is None else self.prolog.query(q, maxresult=maxresult)
            for sol in it:
                yield sol
        finally:
            pass

    from contextlib import contextmanager
    @contextmanager
    def _open_query(self, functor, args):
        q = Query(functor(*args))
        try:
            yield q
        finally:
            q.close()

    def conditional_entailment_prob(dx: AlpDiagnosis, edu, explanation: list[str], source_text: str) -> float:
        """
        P(EDU | H), approximated by NLI/entailment or logic.
        Prolog side: p_edu_given_h(ClaimAtom, HList, P).
        edu.claim_atom is used as the 'hypothesis' to be entailed under H.
        """
        if not explanation:
            return 1e-3

        h_list = "[" + ", ".join(explanation) + "]"
        query = f"p_edu_given_h({edu.claim_atom}, {h_list}, P)"

        results = list(dx.prolog.query(query, maxresult=1))
        if not results:
            return 1e-3
        return float(results[0]["P"])

    # ---------- PLP probability: plp_prob and plp_prob_single ----------

    def plp_prob(dx: AlpDiagnosis, explanation: list[str]) -> float:
        """
        Probabilistic logic program score P(H).
        Prolog side: plp_prob(HList, P).
        """
        if not explanation:
            return 1.0

        h_list = "[" + ", ".join(explanation) + "]"
        query = f"plp_prob({h_list}, P)"

        results = list(dx.prolog.query(query, maxresult=1))
        if not results:
            return 1e-6
        return float(results[0]["P"])

    def plp_prob_single(dx: AlpDiagnosis, h: str) -> float:
        """
        Probability of a single hypothesis literal h.
        Prolog side: plp_prob_single(H, P).
        """
        query = f"plp_prob_single({h}, P)"
        results = list(dx.prolog.query(query, maxresult=1))
        if not results:
            return 1e-6
        return float(results[0]["P"])

if __name__ == "__main__":
    dx = AlpDiagnosis()
    dx.load_demo()

    # Gout-like case
    s1 = ["severe_joint_pain", "swelling", "uric_acid_high", "sudden_onset_night"]
    print("Symptoms:", s1)
    print("Explanation:", dx.explain_obs(s1))  # -> ['disease(gout)']

    # RA-like case
    s2 = ["symmetric_small_joints_pain", "morning_stiffness_gt_60min", "swelling", "elevated_esr_crp"]
    print("Symptoms:", s2)
    print("Explanation:", dx.explain_obs(s2))  # -> ['disease(rheumatoid_arthritis)']
